File: backend/app/models/card.py
"""Card — a flashcard with SM-2 spaced-repetition state.

Mirrors ios/Cram/Models/Card.swift. The SM-2 fields (ease_factor, interval_days,
repetitions, lapses) are the canonical state; due_date is the effective, possibly
exam-compressed next-review date (ADR 0002 / ADR 0004).
"""

from __future__ import annotations

import uuid
from datetime import datetime
from typing import TYPE_CHECKING

from sqlalchemy import DateTime, Double, ForeignKey, Integer, String, Text, func
from sqlalchemy.orm import Mapped, mapped_column, relationship

from ..db import Base
from .mixins import OwnedMixin, SyncMixin, UUIDPkMixin

if TYPE_CHECKING:
    from .review_log import ReviewLog
    from .source import Source
    from .subject import Subject


class Card(UUIDPkMixin, OwnedMixin, SyncMixin, Base):
    __tablename__ = "cards"

    subject_id: Mapped[uuid.UUID] = mapped_column(
        ForeignKey("subjects.id", ondelete="CASCADE"), nullable=False, index=True
    )
    # Cards outlive the source they came from; null the link rather than delete the card.
    source_id: Mapped[uuid.UUID | None] = mapped_column(
        ForeignKey("sources.id", ondelete="SET NULL"), nullable=True, index=True
    )
    front: Mapped[str] = mapped_column(Text, nullable=False)
    back: Mapped[str] = mapped_column(Text, nullable=False)
    topic: Mapped[str] = mapped_column(String(256), nullable=False)
    # Author-estimated difficulty, 1 (easy) … 5 (hard); the range is enforced by the API layer.
    difficulty: Mapped[int] = mapped_column(Integer, nullable=False, default=3)

    # SM-2 canonical state.
    ease_factor: Mapped[float] = mapped_column(Double, nullable=False, default=2.5)
    interval_days: Mapped[int] = mapped_column(Integer, nullable=False, default=0)
    repetitions: Mapped[int] = mapped_column(Integer, nullable=False, default=0)
    lapses: Mapped[int] = mapped_column(Integer, nullable=False, default=0)
    # New cards are due immediately.
    due_date: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), server_default=func.now(), nullable=False, index=True
    )

    subject: Mapped["Subject"] = relationship(back_populates="cards")
    source: Mapped["Source | None"] = relationship(back_populates="cards")
    review_logs: Mapped[list["ReviewLog"]] = relationship(
        back_populates="card", cascade="all, delete-orphan", passive_deletes=True
    )

File: ios/Cram/Models/Card.swift
import Foundation
import SwiftData

/// A flashcard with SM-2 spaced-repetition state.
///
/// The SM-2 state (`easeFactor`, `intervalDays`, `repetitions`, `lapses`) is the canonical
/// source of truth, updated only by the scheduler's SM-2 step. `dueDate` is the *effective*
/// next-review date, which may be compressed toward the subject's exam date (see ADR 0004).
@Model
final class Card {
    @Attribute(.unique) var id: UUID
    var front: String
    var back: String
    var topic: String
    /// Author-estimated difficulty, 1 (easy) … 5 (hard).
    var difficulty: Int
    var createdAt: Date

    // SM-2 canonical state.
    var easeFactor: Double
    var intervalDays: Int
    var repetitions: Int
    var lapses: Int
    /// Effective next-review date (may be exam-compressed). Surfaced when `dueDate <= now`.
    var dueDate: Date

    var subject: Subject?
    var source: Source?

    // Sync metadata (v0.5 Phase 5). The SM-2 state above is what changes on review; bump
    // `updatedAt` / `needsSync` via `touch()` whenever the scheduler rewrites it.
    var updatedAt: Date = Date()
    var deletedAt: Date?
    var needsSync: Bool = true

    @Relationship(deleteRule: .cascade, inverse: \ReviewLog.card)
    var reviewLogs: [ReviewLog] = []

    init(front: String,
         back: String,
         topic: String,
         difficulty: Int = 3,
         subject: Subject? = nil,
         source: Source? = nil) {
        self.id = UUID()
        self.front = front
        self.back = back
        self.topic = topic
        self.difficulty = min(max(difficulty, 1), 5)
        self.createdAt = .now
        self.easeFactor = 2.5      // SM-2 default
        self.intervalDays = 0
        self.repetitions = 0
        self.lapses = 0
        self.dueDate = .now        // brand-new cards are due immediately
        self.subject = subject
        self.source = source
        self.updatedAt = .now
        self.needsSync = true
    }

    /// Whether this card is due for review at the given moment.
    func isDue(asOf date: Date = .now) -> Bool { dueDate <= date }

    /// 0…1 mastery estimate from SM-2 state (1 = well known), for scheduling weight & progress.
    var mastery: Double {
        // Ease ranges ~1.3…2.5+; combine with repetition count for a rough strength signal.
        let easeComponent = min(max((easeFactor - 1.3) / (2.5 - 1.3), 0), 1)
        let repComponent = min(Double(repetitions) / 5.0, 1)
        return 0.5 * easeComponent + 0.5 * repComponent
    }
}
